### DISCUSSION TITLE
Reorganizing the docs for Graphcache

### DISCUSSION THREAD

**Discussion by icecapp** [urql-graphql/urql#1865] [Q&A] [answered]
Unpopular opinion alert!

urql is generally lauded for having fantastic docs. I'm an experienced dev, relatively new to GraphQL and unfortunately I find the docs outstandingly difficult to understand and navigate. (I do find them better than Apollo's tho)

Having started with urql a few weeks ago, whenever I try to figure out how to do something, I find myself presented with an unreasonably large text/code ratio. Instead of just getting you started and pointing to more in-depth docs for further reading, it presents you with in-depth explanations first, doesn't give you a succinct path to getting started, and tries to teach you about the inner workings of urql, which frankly, as a user, I really should not/do not care about to get started. 

The section of the docs that have been the clearest to me so far is Offline Support, which I haven't gotten to use yet, because the Graphcache docs it builds on are basically unapproachable for me. 80% of the first section of the Graphcache docs `Normalized Caching` doesn't even seem to be urql-specific, it reads like a high level (yet super long) explanation of what a normalized cache is and how it works _in general_, in lib-specific-terminology, and the text/code ratio is just too high. Why is an in-depth explanation the intro of a section? 

I read it a dozen times and still don't understand it. Some parts less than others, but e.g. `Storing Normalized Data` has been to me so far, entirely incomprehensible. Here's the first sentence:

> In our case we store all values of fields in a dictionary of their primary key, generated from an ID or other key and type name, and the field’s name and arguments, if it has any.

The more I read this, the less it makes sense. The primary key is generated from an ID and type name, as well as the field's name and arguments? (also "dictionary"? isn't this js?) I don't understand this, but should even I care about this to get started with normalized cache (non-rhetorical question)?

The Normalized Cache docs should just get to the point of how to _use_ the Normalized Cache, _before_ going in depth about how it works. To be more specific, instead of first explaining corner-case stuff like `Custom Keys and Non-Keyable Entities` (I'm willing to bet 90% of entities in the wild have `id` or `_id` as the pk _anyways_.) it should immediately answer practical questions like:

"If I run a query now with Graphcache default setup, will it get cached?" (after Setup and Installation)
"Will running the same query again fetch results from cache?"
"How does cache get invalidated?"
"If I send a mutation, does the response update the cache automatically? If not, how to do so manually?"

In a similar fashion, the auth docs I find beat around the bush _ruthlessly_, until you discover the `Read more about @urql/exchange-auth's API in our API docs` link in the middle of the page, which, when opened, all the way at the end, links to `Read more examples in the documentation given here`, which, wait for it, is a `Quick Start Guide`!!! With the Quick Start Guide in hand, I was able to go _back_ to the docs and gain a thorough understanding of how auth works in urql. 
Without this arguably hidden Quick Start Guide, one has exactly _no way_ to figure out how to go proceed after having completed `Installation & Setup` for Auth (try it!). 

If there were `Quick Start Guide`s like this for most of the sections in the docs, the library would instantly become _so much more approachable_.

Otherwise I would love to make more structural suggestions about the docs in the form of PRs but unfortunately I do not fully understand how urql works yet (specifically graphcache). Once I wrap my head around it, I will.

How did you end up learning how to use graphcache ?

**Comment by kitten** [comment, marked as answer]
Hiya 👋 First, thanks for the input! I really appreciate it and would never say any of it should come across as unpopular. From a technical writing perspective, urql has been one of my biggest personal challenges so far. I obviously don't speak for everyone who has worked on the docs, but I believe it's a delicate problem of continuous iteration.

In general, what I'm thinking of when writing them is:

1. What is the order of information that decreases the chance of confusion later on (I know, that sounds quite pedagogic, ha)
2. The ratio of glanceable/skimmable content to in-depth information to deliver the right bits at the right moment? (In other words, not quite a balance of text:code but explanation:instruction)
3. How much knowledge is needed to "counter" or complement existing intuitions and preconceptions a reader may have

This explains why I chose to write the "Normalised Caching" page how it is. In fact, a "Basics Graphcache" page doesn't exist and is closer to the following two pages.

The first page exists because I came across many misunderstandings when it comes to how people think about normalised caching and GraphQL. Some people have found it useful to have a mental model of what goes on behind the scenes, but the hardest thing to later on do is condensing documentation 😃 In general, hierarchy, touching points, and links between pages are prioritsed over condensing content. Hence the information ratio varies wildly and you'll find pages growing until they're rewritten at which point they often shrink.
I believe that Graphcache's docs are currently past a "growth phase" and need condensing.

Similarly, the Auth docs haven't been revisited for a while and require some attention.

Currently, my next focus is on improving the mention of "context" on operations, which is glanced over a lot in the docs: https://github.com/FormidableLabs/urql/pull/1800#issuecomment-881539236

However, I do think most of the questions you mentioned are answered on the "Normalised Caching" page (although invalidation is a longer topic to cover and tricky to answer, so it's split out into updates)

But I think overall, the page indeed doesn't do enough to indicate that the underlying principles in the first section are "skimmable" by glancing over some details and takes too long to redirect details to the "resolvers" and "updates" pages that follow, since the mental model splits Graphcache into "customising querying" and "customising updates".

I'm not quite willing to skip over keying and the nornalisation process entirely however. The most common issue I'm aware of of learners (especially those coming from Apollo) is confusion about how the data is keyed and stored and what it means when data is unkeyed/unkeyable.

Would you agree with the assessment though that this page mostly needs some compaction?

The other problem to address is likely that most other pages go about things by starting off with "Get Started like this" while Graphcache's subsection is more like a meta-page. Would you say that's what may have thrown you off?

**Reply by kitten** [reply]
I've created a quick project board, just to capture what I see are the most pressing changes I'd like to make to the docs https://github.com/FormidableLabs/urql/projects/6

After taking a look at this and my thoughts and comparing everything against Graphcache's docs, I think you're 100% right. They're not fit for purpose and need updating more urgently than I thought. Generally, I think the biggest problem is that they break expectations and we're also introducing normalised caching quite late. Graphcache itself should be hinted at early like "Document Caching" is, without much of a fanfare.

Its first page should be condensed and replaced with "setup-focused" information and cut out all superfluous information. It's obviously not all superfluous, but it isn't elegant in how it orders information and instead "lists it all" instead of focusing on what's relevant first.

**Reply by GollyJer** [reply]
Hey @kitten. Just wanted to chime in. I read through the todo's you created and they all look great. Making the transition to graphcache easier to understand will be super helpful.

**Comment by icecapp** [comment]
Thanks for the quick answer! That was _fast_.

>Would you agree with the assessment though that this page mostly needs some compaction?

To be fair I've yet to fully wrap my head around it, so it's hard to say if it's length, clarity, organization...

> But I think overall, the page indeed doesn't do enough to indicate that the underlying principles in the first section are "skimmable" by glancing over some details and takes too long to redirect details to the "resolvers" and "updates" pages that follow, since the mental model splits Graphcache into "customising querying" and "customising updates".

Yes

> Generally, I think the biggest problem is that they [Graphcache docs] break expectations

On the main "Graphcache" page it also states:
> Graphcache already does what you'd expect it to do without any additional configuration.

I for one do not know what behavior to expect from a normalized cache. Like by default, when does caching occur? I found hints here and there such as:

> When the mutation is executed and we get the result back, Graphcache already writes the Todo item to its normalized cache.

(Ok, so... mutation responses are autocached ?)

> The other problem to address is likely that most other pages go about things by starting off with "Get Started like this" while Graphcache's subsection is more like a meta-page. Would you say that's what may have thrown you off?

That's a very low hanging fruit imho yes. A proper Quick Start Guide like the one that exists for Auth would demystify the feature. 

 FWIW I actually asked for advice on getting started with urql with a link to the graphcache page on [Reactiflux](https://discord.com/channels/102860784329052160/103315563073269760/876901001012060201) and someone voiced the exact sentiment.

**Comment by icecapp** [comment]
Hi again! So I was working on another part of my project in the meantime, but as of now, my graphql client is on the critical path. 

I am 100% willing to help with the Normalized Cache docs if someone can explain how Graphcache works. I just reread the entire Normalized Cache section _twice_ with not much further understanding. The most pressing clarification required IMHO is:

> in a lot of cases, Graphcache already does what you'd expect it to do without any additional configuration.

What _does_ Graphcache do by default? I haven't used Apollo, as mentioned in the original post, so I literally have no functional expectations from a normalized cache/graphcache, asides from the fact that query results get stored in a normalized cache. Is that all it does? If so we should explicitly say so. If not, I suggest listing out everything it does right after the quote above (something like: What you get out-of-the-box:) If someone can explain this to me, I can write it up.

PS I have a bunch of other questions and suggestions but I figure tackling one at a time respects people's time/cognitive load.

**Reply by kitten** [reply]
I do have some more explanations in this post:
https://kitten.sh/graphql-normalized-caching

> What does Graphcache do by default?

It caches everything by default by keys, so it builds up a big map of relations and objects. It's updates and invalidations (well, and custom, computed resolvers) that are customised as needed.

**Comment by icecapp** [comment]
deleted previous reply as became irrelevant

Hi there, I read that and still have a hard time wrapping my head around things. I figured if I just tried putting it to practice I'd uncover the big picture as I go. 

Here's a question I couldn't easily find an answer for:

How do you fetch a record from the cache by ID/arg?

**Reply by manojsethi** [reply]
Hi @icecapp 

I guess we both are on same page. Please have a look at this answer explaining what to use and when

https://github.com/FormidableLabs/urql/discussions/1918#discussioncomment-1276587

Regarding fetching of a record with id you can simply use the cache.readQuery if record you want to fetch was part of query you ui has executed. So with this case a complete result will be returned back.

If you want to get a particular field then you can utilize the power of cache.resolve with combination of cache.keyOfEntity.

I am also learning so a thumbs up on this from @kitten  @JoviDeCroock would be helpful.

**Reply by icecapp** [reply]
So I was told this is not the "graphql normalized cache" way of handling things. 

You'd want to just make the relevant query as if it were outgoing to the server, and let the cache handle it.
